govcd/nsxv_ipset.go

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"encoding/xml"
+	"fmt"
+	"net/http"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+func (vdc *Vdc) CreateNsxvIpSet(ipSetConfig *types.EdgeIpSet) (*types.EdgeIpSet, error) {
+	if err := validateCreateNsxvIpSet(ipSetConfig); err != nil {
+		return nil, err
+	}
+
+	vdcId, err := getUuidFromHref(vdc.Vdc.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get vdc ID from HREF: %s", err)
+	}
+
+	// build a path for IP set creation. The endpoint should look like:
+	// https://_hostname_/network/services/ipset/f9daf2da-b4f9-4921-a2f4-d77a943a381c where the
+	// trailing UUID is vDC ID
+	httpPath, err := vdc.buildNsxvNetworkServiceEndpointURL(types.NsxvIpSetServicePath + "/" + vdcId)
+	if err != nil {
+		return nil, fmt.Errorf("could not get network services API endpoint for IP set: %s", err)
+	}
+
+	// Success or an error of type types.NSXError is expected
+	_, err = vdc.client.ExecuteParamRequestWithCustomError(httpPath, nil, http.MethodPost, types.AnyXMLMime,
+		"error creating IP set: %s", ipSetConfig, &types.NSXError{})
+	if err != nil {
+		return nil, err
+	}
+
+	createdIpSet, err := vdc.GetNsxvIpSetByName(ipSetConfig.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup newly created IP set with name %s: %s", ipSetConfig.Name, err)
+	}
+
+	return createdIpSet, nil
+}
+
+// UpdateNsxvIpSet sends all fields of ipSetConfig. Omiting a value may reset it. ID is mandatory to
+// perform update.
+// Because the API always requires a Revision to be sent - the update fetches latest revision number
+// automatically and embeds into the update structure.
+func (vdc *Vdc) UpdateNsxvIpSet(ipSetConfig *types.EdgeIpSet) (*types.EdgeIpSet, error) {
+	err := validateUpdateNsxvIpSet(ipSetConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inject latest Revision for this IP set so that API accepts change
+	currentIpSet, err := vdc.GetNsxvIpSetById(ipSetConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch current IP set: %s", err)
+	}
+	ipSetConfig.Revision = currentIpSet.Revision
+
+	httpPath, err := vdc.buildNsxvNetworkServiceEndpointURL(types.NsxvIpSetServicePath + "/" + ipSetConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get network services API endpoint for IP set: %s", err)
+	}
+
+	// Result is either 204 for success, or an error of type types.NSXError
+	errString := fmt.Sprintf("error while updating IP set with ID %s :%%s", ipSetConfig.ID)
+	_, err = vdc.client.ExecuteRequestWithCustomError(httpPath, http.MethodPut, types.AnyXMLMime,
+		errString, ipSetConfig, &types.NSXError{})
+	if err != nil {
+		return nil, err
```

Comment:
Custom message on what action error happen would be helpfull

Reply:
same here, error message in line 72